### ISSUE TITLE
Memory leak in ngs::ReadCollection and/or ncbi::NGS::openReadCollection when streaming SRA records?

### ISSUE THREAD

**Issue by jgans** [ncbi/ncbi-vdb#32]
There appears to be a memory leak in ngs::ReadCollection and/or ncbi::NGS::openReadCollection when **streaming** SRA records (as opposed to reading SRA records that have been previously downloaded using the prefetch utility). This leak can be observed using the following code example:
```
#include <iostream>
#include <sstream>
#include <sys/types.h> // For getpid()
#include <unistd.h>
#include <ncbi-vdb/NGS.hpp> // For openReadCollection

using namespace std;

int main(int argc, char *argv[])
{
        const string accession = "SRR360112";

        cout << "Starting memory test for " << accession << endl;

        const pid_t pid = getpid();

        for(size_t iteration = 0;iteration < 100;++iteration){

                stringstream mem_check;
                // Very simple way to monitor memory usage on Linux
                mem_check << "cat /proc/" << pid << "/status | grep VmSize";

                system( mem_check.str().c_str() );

                // Open the SRA accession, but don't load any data ...
                ngs::ReadCollection run(  ncbi::NGS::openReadCollection(accession) );
        }

        return EXIT_SUCCESS;
}
```

Running this simple code on an AWS instance that only _opens_ the SRA record, and does not actually load any data, yields the following memory consumption:
<img width="533" alt="image" src="https://user-images.githubusercontent.com/36452131/92146642-3e779f80-edd7-11ea-9c59-1e49b86208d1.png">

The growing memory consumption is observed for many SRA records, not just the specific example of SRR360112 that is used above.

If the SRA record is not already stored locally (i.e. by first running the prefetch utility), the memory consumption grows after every iteration. This growing memory consumption becomes a problem when loading data from thousands of SRA records (especially when using AWS EC2 instances for which extra RAM means extra $$$).

**Comment by jgans** [comment]
Digging around in the ncbi-vdb code with the Valgrind memory tool, it appears that some of the leaking memory is due to a `KDataBuffer` object not being released in the `KClientHttpRequestSendReceiveNoBodyInt()` function. Adding a few calls to `KDataBufferWhack` to the `KClientHttpRequestSendReceiveNoBodyInt()` function in file `ncbi-vdb/libs/kns/http-request.c` reduces, but does not eliminate, the memory leak:
```
--- http-request.c.orig 2020-09-05 20:42:15.753732694 +0000
+++ http-request.c      2020-09-05 20:50:44.294992766 +0000
@@ -1481,27 +1481,36 @@
 
         /* create message */
         rc = KClientHttpRequestFormatMsgInt ( self, & buffer, method, uriForm );
-        if ( rc != 0 )
+        if ( rc != 0 ){
+
+           KDataBufferWhack( & buffer );
             break;
+       }
 
         /* send the message and create a response */
         rc = KClientHttpSendReceiveMsg ( self -> http, _rslt,
             (char *) buffer.base,
             buffer.elem_count - 1, NULL, (char *) self -> url_buffer . base );
+
         if ( rc != 0 )
         {
             KClientHttpClose ( self -> http );
             rc = KClientHttpSendReceiveMsg ( self -> http, _rslt,
                 (char *) buffer.base, buffer.elem_count - 1, NULL,
                 (char *) self -> url_buffer . base );
-            if ( rc != 0 )
+            if ( rc != 0 ){
+
+               KDataBufferWhack( & buffer );
                 break;
+           }
         }
 
         rslt = * _rslt;
         rslt -> expiration = expiration; /* expiration has to reach the caller */
         expiration = NULL;
 
+       KDataBufferWhack( & buffer );
```

Valgrind is still reporting a memory leak:
```
==15342== 4,224 (112 direct, 4,112 indirect) bytes in 1 blocks are definitely lost in loss record 87 of 103
==15342==    at 0x4C31B25: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15342==    by 0x1EC6FB: KClientHttpAddHeaderString.part.1 
==15342==    by 0x1ED3E9: KClientHttpVAddHeader
==15342==    by 0x1ED47C: KClientHttpAddHeader
==15342==    by 0x1F30A1: KClientHttpRequestPOST_Int 
==15342==    by 0x1F3374: KClientHttpRequestPOST
==15342==    by 0x1F3574: KClientHttpRequestHEAD
==15342==    by 0x1EB811: KNSManagerVMakeHttpFileIntUnstableImpl
==15342==    by 0x1EC560: KNSManagerVMakeHttpFileIntUnstableFromBuffer
==15342==    by 0x1E5B06: KNSManagerMakeReliableHttpFile
==15342==    by 0x1AC46F: VFSManagerMakeHTTPFile
==15342==    by 0x1AD227: VFSManagerOpenCurlFile.part.6
```
but I haven't been able to figure this one out yet ...

**Comment by boshkins** [comment]
@jgans thank you for the report and the investigation! We have opened an internal ticket and will fix this soon.

**Comment by klymenko** [comment]
The leak was fixed in 2.10.9 release.
